controllers/apim/virtualservice_controller.go

```diff
@@ -0,0 +1,218 @@
+		logger.Error(err, "failed to add owner ref to AuthorizationPolicy resource")
+		return err
+	}
+	err := r.ReconcileResource(ctx, &istiosecurityv1beta1.AuthorizationPolicy{}, &authPolicy, alwaysUpdateAuthPolicy)
+	if err != nil && !apierrors.IsAlreadyExists(err) {
+		logger.Error(err, "ReconcileResource failed to create/update AuthorizationPolicy resource")
+		return err
+	}
+
+	logger.Info("successfully created/updated AuthorizationPolicy resource(s)")
+	return nil
+}
+
+func alwaysUpdateAuthPolicy(existingObj, desiredObj client.Object) (bool, error) {
+	existing, ok := existingObj.(*istiosecurityv1beta1.AuthorizationPolicy)
+	if !ok {
+		return false, fmt.Errorf("%T is not a *istiosecurityv1beta1.AuthorizationPolicy", existingObj)
+	}
+	desired, ok := desiredObj.(*istiosecurityv1beta1.AuthorizationPolicy)
+	if !ok {
+		return false, fmt.Errorf("%T is not a *istiosecurityv1beta1.AuthorizationPolicy", desiredObj)
+	}
+
+	existing.Spec = desired.Spec
+	return true, nil
+}
+
+func normalizeStringMatch(sm *networkingv1alpha3.StringMatch) string {
+	if prefix := sm.GetPrefix(); prefix != "" {
+		return prefix + "*"
+	}
+	if exact := sm.GetExact(); exact != "" {
+		return exact
+	}
+	// Regex string match is not supported because authpolicy doesn't as well.
+	return ""
+}
+
+// VirtualServiceFilter allows generation of only relevant reconciliation request.
+type VirtualServiceFilter struct {
+	K8sClient client.Client
+	Logger    logr.Logger
+}
+
+// Map contains filtering logic for virtualservice.
+func (m *VirtualServiceFilter) Map(obj client.Object) []reconcile.Request {
+	virtualServiceAnnotations := obj.GetAnnotations()
+	_, present := virtualServiceAnnotations[KuadrantAuthProviderAnnotation]
+	if !present {
+		authObjKey := types.NamespacedName{
+			Name:      getAuthPolicyName(obj.GetName()),
+			Namespace: obj.GetNamespace(),
+		}
+
+		var authPolicy client.Object
+		if err := m.K8sClient.Get(context.Background(), authObjKey, authPolicy); err != nil {
+			// no annotation but authpolicy exist means annotation was removed
+			if !apierrors.IsNotFound(err) {
+				m.Logger.Error(err, "failed to check AuthorizationPolicy existence")
+			}
+			return nil // this virtualservice is not protected
+		}
+
+		// AuthorizationPolicy exists
+		if err := m.K8sClient.Delete(context.Background(), authPolicy); err != nil {
+			m.Logger.Error(err, "failed to delete orphan authorizationpolicy")
+		}
+		return nil
+	}
+
+	return []reconcile.Request{
+		{NamespacedName: types.NamespacedName{
+			Name:      obj.GetName(),
+			Namespace: obj.GetNamespace(),
+		}},
+	}
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *VirtualServiceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	virtualServiceFilter := VirtualServiceFilter{
+		K8sClient: r.Client(),
+		Logger:    r.Logger().WithName("AuthPolicyHandler"),
+	}
+	return ctrl.NewControllerManagedBy(mgr).
+		Watches(
```

Comment:
Check the services controller of the current main branch. Use a label instead of annotation for builtin filtering out of events. https://github.com/Kuadrant/kuadrant-controller/blob/main/controllers/service_controller.go#L341

```go
return ctrl.NewControllerManagedBy(mgr).                                                                   
    For(&corev1.Service{}, builder.WithPredicates(kuadrantServicesPredicate())).                           
    Owns(&networkingv1beta1.API{}).                                                                        
    WithLogger(log.Log). // use base logger, the manager will add prefixes for watched sources             
    Complete(r) 
```

Reply:
`For` and `Owns`, confused me a bit. But it makes sense since AuthorizationPolicy is being generated. But I am not sure we need that `WithPredicated` since filtering is the same for all the events.

Reply:
Thanks, @eguzki! looks much more simpler now with `For` and `Owns` :)